[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py

coconutruben opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #156477
* #156292
* #156282
* __->__ #156281
* #156280
* #156279
* #156278
* #156277
* #156276

Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

coconutruben committed:
[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py

Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py"


Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py"


Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py"


Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py"


Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben commented:
failure seems related to me moving `GeneratedCodeCache` out of select_algorithm. I'll fix and rerun

```
FAILED [0.0029s] inductor/test_max_autotune.py::TestMaxAutotune::test_triton_template_generated_code_cache_key - AttributeError: module 'torch._inductor.select_algorithm' has no attribute 'GeneratedCodeCache'
```

masnesral reviewed (changes requested):
Test failure looks relevant: 

```
FAILED [0.0029s] inductor/test_max_autotune.py::TestMaxAutotune::test_triton_template_generated_code_cache_key - AttributeError: module 'torch._inductor.select_algorithm' has no attribute 'GeneratedCodeCache'
```

coconutruben committed:
Update on "[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py"


Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor][2/2] break out TritonTemplate, TritonTemplateKernel, TritonTemplateCaller out of select_algorithm.py"


Summary:
\# Why

- simplify import and dependency structure
- simplify select_algorithm.py and keep semantically closer to actually select algorithm
- consolidate all triton template related things in triton_template/

\# What

This is the second part of a 2 part PR as the whole thing is too large.
The second part now removes the old implementations of those that are
already copied over, and points them to the new call sites.

- pull out out select_algorithm.py
  - TritonTemplate
  - TritonTemplateCaller
  - TritonTemplateKernel
and move into triton_heuristics

- copy SymbolicGridFn into triton_templates/common.py

The copying here is done to simplify (I know) the PR and later migrate other users when they migrate into triton_templates/

- fix import sites across

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.xrhwr0nuqveh

Test Plan:
- ci testing

```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

```
V0617 19:48:23.589000 3601502 site-packages/torch/_inductor/select_algorithm.py:609] [0/0] Precompilation elapsed time: 0.00s
V0617 19:48:23.590000 3601502 site-packages/torch/_inductor/select_algorithm.py:638] [0/0] Autotuning elapsed time: 0.00s
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
